src/integrator/cuda/cuda_pack_density.cu

```diff
@@ -7,54 +10,84 @@ namespace cuda       {
   const auto  LDAS              = task.nbe;
         auto* ASmall_device     = task.nbe_scr;
 
   //if( LDAS == LDAB ) return;
 
-
-  const int tid_x = blockDim.x * blockIdx.x + threadIdx.x;
-  const int tid_y = blockDim.y * blockIdx.y + threadIdx.y;
-
-  int64_t i(0);
-  for( size_t i_cut = 0; i_cut < ncut; ++i_cut ) {
-    const int64_t i_cut_first  = submat_cut_device[ 2*i_cut ];
-    const int64_t i_cut_second = submat_cut_device[ 2*i_cut + 1 ];
-    const int64_t delta_i      = i_cut_second - i_cut_first;
-
-    int64_t j(0);
-  for( size_t j_cut = 0; j_cut < ncut; ++j_cut ) {
-    const int64_t j_cut_first  = submat_cut_device[ 2*j_cut ];
-    const int64_t j_cut_second = submat_cut_device[ 2*j_cut + 1 ];
-    const int64_t delta_j      = j_cut_second - j_cut_first;
-
-    auto* ASmall_begin = ASmall_device + i           + j          *LDAS;
-    auto* ABig_begin   = A             + i_cut_first + j_cut_first*LDA ;
-    
-    for( size_t J = tid_y; J < delta_j; J += blockDim.y )      
-    for( size_t I = tid_x; I < delta_i; I += blockDim.x )
-      ASmall_begin[I + J*LDAS] = ABig_begin[I + J*LDA];
-
-    j += delta_j;
+  const int tid_xx = threadIdx.x % WARP_X;
+  const int tid_xy = threadIdx.x / WARP_X;
+
+  const int tid_yx = threadIdx.y % CUT_X;
+  const int tid_yy = threadIdx.y / CUT_X;
+
+  const int start_cut_y = submat_block_device[block_y];
+  const int end_cut_y   = submat_block_device[block_y+1];
+  const int start_cut_x = submat_block_device[block_x];
+  const int end_cut_x   = submat_block_device[block_x+1];
+
+  for( int i_cut = tid_yy + start_cut_y; i_cut < end_cut_y; i_cut += CUT_Y ) {
+    const int i_cut_first  = submat_cut_device[ 4*i_cut ];
+    const int delta_i      = submat_cut_device[ 4*i_cut + 1 ];
+    const int i_cut_small  = submat_cut_device[ 4*i_cut + 2 ];
+
+  for( int j_cut = tid_yx + start_cut_x; j_cut < end_cut_x; j_cut += CUT_X ) {
+    const int j_cut_first  = submat_cut_device[ 4*j_cut ];
+    const int delta_j      = submat_cut_device[ 4*j_cut + 1 ];
+    const int j_cut_small  = submat_cut_device[ 4*j_cut + 2 ];
+
+    auto* ASmall_begin = ASmall_device + i_cut_small + j_cut_small*LDAS;
+    auto* ABig_begin   = A   + i_cut_first + j_cut_first*LDA;
+
+    int J;
+    for( J = tid_xy; J < (delta_j / EFF_UNROLL) * EFF_UNROLL; J += EFF_UNROLL ) {
+      for( int I = tid_xx; I < delta_i; I += WARP_X ) {
+
+        double val[UNROLL_FACTOR];
+        double* address[UNROLL_FACTOR];
+#pragma unroll
+        for (int k = 0; k < UNROLL_FACTOR; k++) {
+          val[k] = ABig_begin[I + (J + k*WARP_Y)*LDA];
+          address[k] = ASmall_begin + I + (J + k*WARP_Y) * LDAS;
+        }
+#pragma unroll
+        for (int k = 0; k < UNROLL_FACTOR; k++) {
+	  // Suggest that the result be evicted first. CUDA11 provides
+          // a non-PTX way to do this so we can update if we require CUDA11
+          asm ("st.global.cs.f64 [%0], %1;" :: "l"(address[k]), "d"(val[k]));
```

Comment:
In CUDA 11.0+ we can replace the inline-PTX with intrinsic functions.
https://docs.nvidia.com/cuda/archive/11.0/cuda-c-programming-guide/index.html#stx-functions

Reply:
I'll have to look into whether or not they have full support for CUDA 11 on Summit (they have a module, but iirc its experimental viz a few features). We can always just put this in a CPP check for CUDA version though (then deprecate once CUDA 11 is widely adopted @ HPC centers)